src/test/v1.2/advanced/channels-support-1.2.ts

```diff
@@ -0,0 +1,272 @@
+
+
+declare let fdc3: DesktopAgent;
+
+let listener1: Listener, listener2: Listener
+
+export class ChannelControl1_2 implements ChannelControl<Channel, Context> {
+  retrieveAndJoinChannel = async (
+    channelNumber: number
+  ): Promise<Channel> => {
+    const channel = await this.getUserChannel(channelNumber);
+    await fdc3.joinChannel(channel.id);
+    return channel;
+  };
+
+  getSystemChannels = async () => {
+    return await fdc3.getSystemChannels();
+  }
+
+  leaveChannel = async () => {
+    return await fdc3.leaveCurrentChannel();
+  }
+
+  getUserChannel = async (channel: number): Promise<Channel> => {
+    const channels = await fdc3.getSystemChannels();
+    if (channels.length > 0) {
+      return channels[channel - 1];
+    } else {
+      assert.fail("No system channels available for app A");
+    }
+  };
+
+  joinChannel = async (channel: Channel): Promise<void> => {
+    return fdc3.joinChannel(channel.id)
+  }
+
+  createTestChannel = async (name: string = "test-channel"): Promise<Channel> => {
+    return fdc3.getOrCreateChannel(name);
+  }
+
+  unsubscribeListeners = (): void => {
+    if (listener1 !== undefined) {
+      listener1.unsubscribe();
+      listener1 = undefined;
+    }
+
+    if (listener2 !== undefined) {
+      listener2.unsubscribe();
+      listener2 = undefined;
+    }
+  }
+
+  channelCleanUp = async (): Promise<void> => {
+    this.unsubscribeListeners();
+    await fdc3.leaveCurrentChannel();
+  }
+
+
+  closeChannelsAppWindow = async (testId: string): Promise<void> => {
+    //Tell ChannelsApp to close window
+    const appControlChannel = await broadcastAppChannelCloseWindow(testId);
+
+    //Wait for ChannelsApp to respond
+    await waitForContext("windowClosed", testId, appControlChannel);
+    await wait(constants.WindowCloseWaitTime);
+  }
+
+
+
+  initCompleteListener = async (testId: string): Promise<Context> => {
+    return waitForContext(
+      "executionComplete",
+      testId,
+      await fdc3.getOrCreateChannel("app-control")
+    );
+  }
+
+  openChannelApp = async (testId: string, channelId: string | undefined, commands: string[], historyItems: number = undefined, notify: boolean = true): Promise<void> => {
+    const channelsAppConfig: ChannelsAppConfig = {
+      fdc3ApiVersion: "1.2",
+      testId: testId,
+      userChannelId: channelId,
+      notifyAppAOnCompletion: notify,
+    };
+
+    if (channelId) {
+      channelsAppConfig.userChannelId = channelId;
+    }
+
+    if (historyItems) {
+      channelsAppConfig.historyItems = historyItems;
+    }
+
+    //Open ChannelsApp then execute commands in order
+    await fdc3.open(
+      "ChannelsApp",
+      buildChannelsAppContext(commands, channelsAppConfig)
+    );
+  }
+
+  setupAndValidateListener1 = (channel: Channel, expectedContextType: string, errorMessage: string, onComplete: (ctx: Context) => void): void => {
+    if (channel) {
+      listener1 = channel.addContextListener(null, (context) => {
```

Comment:
to get this to work in Finsemble, I need to make this `async`

Reply:
Shouldn't need it to be... I don't have to use an async handler in either the prerelease version I'm running, nor 8.2 on either channel type... Both these are working for me:

```javascript
fdc3.addContextListener(null, (context) => {console.log("received context: ", context);})
```

```javascript
const channel = await fdc3.getOrCreateChannel("test");
channel.addContextListener(null, (context) => {console.log("received context via test: ", context);})
```

What output/error are you getting?

Reply:
We went through this and determined that the test is actually broken - adding async was somehow shuffle the order of things so it appeared to work/pass but the test is no longer being implemented as defined due to the lack of a typed context listener here